core/src/main/java/com/oath/oak/InternalOakMap.java

```diff
@@ -665,106 +685,98 @@ the absolute value of the version written in lookUp (this way, if the version is
 
             if (logicallyDeleted) {
                 // This is the case where we logically deleted this entry (marked the value as deleted), but someone
                 // reused the entry before we unlinked it. We have the previous value saved in v.
-                return transformer == null ? Result.withFlag(TRUE) : Result.withValue(v);
+                return transformer == null ? ctx.result.withFlag(TRUE) : ctx.result.withValue(v);
             } else {
-                Result<V> removeResult = valueOperator.remove(lookUp.valueSlice, memoryManager, lookUp.version,
+                Result removeResult = valueOperator.remove(ctx, memoryManager,
                         oldValue, transformer);
                 if (removeResult.operationResult == FALSE) {
                     // we didn't succeed to remove the value: it didn't contain oldValue, or was already marked
                     // as deleted by someone else)
-                    return Result.withFlag(FALSE);
+                    return ctx.result.withFlag(FALSE);
                 } else if (removeResult.operationResult == RETRY) {
                     continue;
                 }
                 // we have marked this value as deleted (successful remove)
                 logicallyDeleted = true;
-                v = removeResult.value;
+                v = (V) removeResult.value;
             }
 
-            assert lookUp.entryIndex > 0;
-            assert lookUp.valueReference != EntrySet.INVALID_VALUE_REFERENCE;
+            assert ctx.entryIndex > 0;
+            assert ctx.value.isAllocated();
 
             // publish
-            if (c.finalizeDeletion(lookUp)) {
+            if (c.finalizeDeletion(ctx)) {
                 rebalance(c);
             }
         }
+
+        throw new RuntimeException("remove failed: reached retry limit (1024).");
     }
 
     OakDetachedBuffer get(K key) {
         if (key == null) {
             throw new NullPointerException();
         }
-        int infiLoopCount = 0;
-        while (true) {
-            infiLoopCount++;
-            if (infiLoopCount > 1000) {
-                System.out.println("Stuck in get");
-                assert false;
-            }
+
+        ThreadContext ctx = getThreadLocalContext();
+
+        for (int i = 0; i < MAX_RETRIES; i++) {
             Chunk<K, V> c = findChunk(key); // find chunk matching key
-            EntrySet.LookUp lookUp = c.lookUp(key);
-            if (lookUp == null || lookUp.valueSlice == null) {
+            c.lookUp(ctx, key);
+            if (!ctx.isValueValid()) {
                 return null;
             }
-            if (updateVersionAfterLinking(c, lookUp)) {
+            if (updateVersionAfterLinking(c, ctx)) {
                 continue;
             }
 
-            return new OakDetachedReadValueBuffer(lookUp.valueReference, lookUp.version, lookUp.keyReference, valueOperator,
-                    memoryManager, this);
+            return getValueReference(ctx);
```

Comment:
Looks like it should be named getValueBuffer() or getValueDetachedBuffer()... Because it doesn't creates reference, at least we doesn't name it reference

Reply:
Fixed.